Upgrade twitter4j (WAS: Dependency review: twitter4j)

### Description

Twitter4j is used in Keycloak for integration with Twitter IDP. This task is to review if twitter4j dependency can be removed entirely.

The point is, that at the time when we added support for Twitter IDP, Twitter did not have support for OAuth 2/ OIDC. However now Twitter has support for it (at least for OAuth2 AFAIK). So this is to doublecheck if we can make Twitter IDP to inherit directly from `OIDCIdentityProvider` or `AbstractOAuth2IdentityProvider` and leverage the existing OAuth2 / OIDC functionality and get rid of twitter4j.

Migration: If we get rid of twitter4j, is it any "migration" needed for the twitter users? For example do they need to update their "application configuration" on Twitter side? If the Twitter integration stop working for existing deployments (without any action needed for users), then we may likely need some "deprecation period" and "migration strategy". But if any update on user's application is not needed, we can likely update our twitter integration right away and get rid of twitter4j.

> Question: If we update directly Twitter provider from "v1" to "v2", will the Twitter login still work for those migrating from previous Keycloak version?

No, it won't. The v1 is using Api Key and Api Key Secret while the v2 needs the Client ID and the Client Secret, which are different data. Although the same app can use v1 or v2 the configuration should be changed from one data set to the other.

> Is it possible to retrieve email somehow by some magic call to some "v1-api-endpoint" even if all the rest stuff will use v2 API?

The v1 API cannot be used with an OAuth2 header. If you call the `verifyCredentials` endpoint with a bearer token generated using OAuth2 the following error is returned:

```
Using OAuth2 in verifyCredentials endpoint:
Caused by: org.keycloak.broker.provider.IdentityBrokerException: failed URL in twitter: {
  "title": "Forbidden",
  "type": "about:blank",
  "status": 403,
  "detail": "You are not permitted to use OAuth2 on this endpoint"
}
```

@pedroigor The different scopes are listed in [this doc page](https://developer.twitter.com/en/docs/authentication/oauth-2-0/authorization-code). Then each v2 API endpoint specifies which scopes are required (for example the `me` endpoint needs `tweet.read` and `users.read`). But I didn't find anyone that returns the email of the authorized account. The `me` endpoint can be called to return more fields but never the email. In v1 `verifyCredential` with email requires a special config option that doesn't even appear in v2. Besides in this [stack-overflow thread](https://stackoverflow.com/questions/74271337/how-to-obtain-users-email-via-twitter-api-v1-or-v2) someone says that is really not provided in v2 and that twitter is (or was) working in providing it using Open ID Connect.

So I suppose that the only option is the third one, providing two providers  (v1 and v2) and deprecate the v1 one. I'll try to also upgrade the twitter4j to the last version in the current v1 prov.

--- a/pom.xml
+++ b/pom.xml
@@ -141,7 +141,7 @@
         <pax.web.version>7.1.0</pax.web.version>
         <servlet.api.30.version>1.0.2.Final</servlet.api.30.version>
         <servlet.api.40.version>2.0.0.Final</servlet.api.40.version>
-        <twitter4j.version>4.0.7</twitter4j.version>
+        <twitter4j.version>4.1.2</twitter4j.version>
         <jna.version>4.1.0</jna.version>
 
         <!-- Databases -->

--- a/services/src/main/java/org/keycloak/social/twitter/TwitterIdentityProvider.java
+++ b/services/src/main/java/org/keycloak/social/twitter/TwitterIdentityProvider.java
@@ -28,6 +28,7 @@ import org.keycloak.broker.provider.IdentityProvider;
 import org.keycloak.broker.provider.util.IdentityBrokerState;
 import org.keycloak.broker.social.SocialIdentityProvider;
 import org.keycloak.common.ClientConnection;
+import org.keycloak.common.util.Base64;
 import org.keycloak.events.Details;
 import org.keycloak.events.EventBuilder;
 import org.keycloak.events.EventType;
@@ -43,12 +44,14 @@ import org.keycloak.services.managers.ClientSessionCode;
 import org.keycloak.services.messages.Messages;
 import org.keycloak.sessions.AuthenticationSessionModel;
 import org.keycloak.vault.VaultStringSecret;
+import twitter4j.AccessToken;
+import twitter4j.OAuthAuthorization;
+import twitter4j.RequestToken;
 import twitter4j.Twitter;
-import twitter4j.TwitterFactory;
-import twitter4j.auth.AccessToken;
-import twitter4j.auth.RequestToken;
-import twitter4j.conf.ConfigurationBuilder;
+import twitter4j.v1.User;
 
+import java.io.ByteArrayInputStream;
+import java.io.ObjectInputStream;
 import javax.ws.rs.GET;
 import javax.ws.rs.QueryParam;
 import javax.ws.rs.WebApplicationException;
@@ -67,14 +70,19 @@ public class TwitterIdentityProvider extends AbstractIdentityProvider<OAuth2Iden
 
     String TWITTER_TOKEN_TYPE="twitter";
 
-
     protected static final Logger logger = Logger.getLogger(TwitterIdentityProvider.class);
 
     private static final String TWITTER_TOKEN = "twitter_token";
-    private static final String TWITTER_TOKENSECRET = "twitter_tokenSecret";
+
+    private final OAuthAuthorization oAuthAuthorization;
 
     public TwitterIdentityProvider(KeycloakSession session, OAuth2IdentityProviderConfig config) {
         super(session, config);
+        try (VaultStringSecret vaultStringSecret = session.vault().getStringSecret(getConfig().getClientSecret())) {
+            oAuthAuthorization = OAuthAuthorization.newBuilder()
+                    .oAuthConsumer(getConfig().getClientId(), vaultStringSecret.get().orElse(getConfig().getClientSecret()))
+                    .build();
+        }
     }
 
     @Override
@@ -84,17 +92,12 @@ public class TwitterIdentityProvider extends AbstractIdentityProvider<OAuth2Iden
 
     @Override
     public Response performLogin(AuthenticationRequest request) {
-        try (VaultStringSecret vaultStringSecret = session.vault().getStringSecret(getConfig().getClientSecret())) {
-            Twitter twitter = new TwitterFactory().getInstance();
-            twitter.setOAuthConsumer(getConfig().getClientId(), vaultStringSecret.get().orElse(getConfig().getClientSecret()));
-
+        try {
             URI uri = new URI(request.getRedirectUri() + "?state=" + request.getState().getEncoded());
-
-            RequestToken requestToken = twitter.getOAuthRequestToken(uri.toString());
+            RequestToken requestToken = oAuthAuthorization.getOAuthRequestToken(uri.toString());
             AuthenticationSessionModel authSession = request.getAuthenticationSession();
 
-            authSession.setAuthNote(TWITTER_TOKEN, requestToken.getToken());
-            authSession.setAuthNote(TWITTER_TOKENSECRET, requestToken.getTokenSecret());
+            authSession.setAuthNote(TWITTER_TOKEN, Base64.encodeObject(requestToken));
 
             URI authenticationUrl = URI.create(requestToken.getAuthenticationURL());
 
@@ -205,16 +208,19 @@ public class TwitterIdentityProvider extends AbstractIdentityProvider<OAuth2Iden
             OAuth2IdentityProviderConfig providerConfig = provider.getConfig();
 
             try (VaultStringSecret vaultStringSecret = session.vault().getStringSecret(providerConfig.getClientSecret())) {
-                Twitter twitter = new TwitterFactory(new ConfigurationBuilder().setIncludeEmailEnabled(true).build()).getInstance();
-                twitter.setOAuthConsumer(providerConfig.getClientId(), vaultStringSecret.get().orElse(providerConfig.getClientSecret()));
-
                 String twitterToken = authSession.getAuthNote(TWITTER_TOKEN);
-                String twitterSecret = authSession.getAuthNote(TWITTER_TOKENSECRET);
+                RequestToken requestToken;
+                try (ObjectInputStream in = new ObjectInputStream(new ByteArrayInputStream(Base64.decode(twitterToken)))) {
+                    requestToken = (RequestToken) in.readObject();
+                }
 
-                RequestToken requestToken = new RequestToken(twitterToken, twitterSecret);
+                AccessToken oAuthAccessToken = provider.oAuthAuthorization.getOAuthAccessToken(requestToken, verifier);
 
-                AccessToken oAuthAccessToken = twitter.getOAuthAccessToken(requestToken, verifier);
-                twitter4j.User twitterUser = twitter.verifyCredentials();
+                Twitter twitter = Twitter.newBuilder()
+                        .oAuthConsumer(providerConfig.getClientId(), vaultStringSecret.get().orElse(providerConfig.getClientSecret()))
+                        .oAuthAccessToken(oAuthAccessToken)
+                        .build();
+                User twitterUser = twitter.v1().users().verifyCredentials();
 
                 BrokeredIdentityContext identity = new BrokeredIdentityContext(Long.toString(twitterUser.getId()));
                 identity.setIdp(provider);

--- a/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/social/TwitterConsentLoginPage.java
+++ b/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/social/TwitterConsentLoginPage.java
@@ -1,0 +1,43 @@
+/*
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.testsuite.pages.social;
+
+import org.jboss.arquillian.graphene.page.Page;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+/**
+ * @author Vaclav Muzikar <vmuzikar@redhat.com>
+ */
+public class TwitterConsentLoginPage extends AbstractSocialLoginPage {
+
+    @FindBy(xpath = "//input[@type='submit' and @id='allow']")
+    private WebElement signInButton;
+
+    @Page
+    TwitterLoginPage loginPage;
+
+    @Override
+    public void login(String user, String password) {
+        // twitter presents a consent page for the application
+        // the SignIn button should be clicked first to go to the real login
+        signInButton.click();
+
+        loginPage.login(user, password);
+    }
+}

--- a/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/social/TwitterLoginPage.java
+++ b/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/social/TwitterLoginPage.java
@@ -17,7 +17,8 @@
 
 package org.keycloak.testsuite.pages.social;
 
-import org.openqa.selenium.NoSuchElementException;
+import org.keycloak.testsuite.util.WaitUtils;
+import org.openqa.selenium.Keys;
 import org.openqa.selenium.WebElement;
 import org.openqa.selenium.support.FindBy;
 
@@ -25,26 +26,25 @@ import org.openqa.selenium.support.FindBy;
  * @author Vaclav Muzikar <vmuzikar@redhat.com>
  */
 public class TwitterLoginPage extends AbstractSocialLoginPage {
-    @FindBy(id = "username_or_email")
+
+    @FindBy(xpath = "//input[@type='text' and @name='text']")
     private WebElement usernameInput;
 
-    @FindBy(id = "password")
+    @FindBy(xpath = "//input[@type='password']")
     private WebElement passwordInput;
-
-    @FindBy(id = "allow")
-    private WebElement loginButton;
 
     @Override
     public void login(String user, String password) {
-        try {
-            usernameInput.clear();
-            usernameInput.sendKeys(user);
-            passwordInput.sendKeys(password);
-        }
-        catch (NoSuchElementException e) { // at some conditions we are already logged in and just need to confirm it
-        }
-        finally {
-            loginButton.click();
-        }
+        // new login page is two phase login (username and then password) and it
+        // needs lots of JS, twitter does not work with default HtmlUnit driver
+        usernameInput.clear();
+        usernameInput.sendKeys(user);
+        usernameInput.sendKeys(Keys.RETURN);
+
+        // wait for the password input to appear
+        WaitUtils.waitUntilElement(passwordInput).is().visible();
+        passwordInput.clear();
+        passwordInput.sendKeys(password);
+        passwordInput.sendKeys(Keys.RETURN);
     }
 }

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/SocialLoginTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/SocialLoginTest.java
@@ -46,7 +46,7 @@ import org.keycloak.testsuite.pages.social.MicrosoftLoginPage;
 import org.keycloak.testsuite.pages.social.OpenShiftLoginPage;
 import org.keycloak.testsuite.pages.social.PayPalLoginPage;
 import org.keycloak.testsuite.pages.social.StackOverflowLoginPage;
-import org.keycloak.testsuite.pages.social.TwitterLoginPage;
+import org.keycloak.testsuite.pages.social.TwitterConsentLoginPage;
 import org.keycloak.testsuite.util.IdentityProviderBuilder;
 import org.keycloak.testsuite.util.OAuthClient;
 import org.keycloak.testsuite.util.RealmBuilder;
@@ -121,7 +121,7 @@ public class SocialLoginTest extends AbstractKeycloakTest {
         FACEBOOK_INCLUDE_BIRTHDAY("facebook", FacebookLoginPage.class),
         GITHUB("github", GitHubLoginPage.class),
         GITHUB_PRIVATE_EMAIL("github", "github-private-email", GitHubLoginPage.class),
-        TWITTER("twitter", TwitterLoginPage.class),
+        TWITTER("twitter", TwitterConsentLoginPage.class),
         LINKEDIN("linkedin", LinkedInLoginPage.class),
         LINKEDIN_WITH_PROJECTION("linkedin", LinkedInLoginPage.class),
         MICROSOFT("microsoft", MicrosoftLoginPage.class),

